llmfoundry/callbacks/async_eval_callback.py

```diff
@@ -0,0 +1,274 @@
+            subset_keys[key] = parameters[key]
+            looking_for.remove(key)
+
+    if looking_for:
+        raise Exception(
+            f'Missing the following required parameters for async eval: {looking_for}\n{parameters}'
+        )
+
+    # Convert the save_folder to a load_path
+    subset_keys['load_path'] = checkpoint
+
+    for logger, config in subset_keys.get('loggers', {}).items():
+        if logger == 'wandb':
+            config['group'] = config.pop('name', training_run_name)
+
+        # mlflow currently does not support grouping, so this will just launch
+        # a new mlflow run
+
+    # Create new eval models list
+    model = subset_keys.pop('model')
+    new_models = {'model_name': model.get('name'), 'model': model}
+
+    tokenizer = subset_keys.pop('tokenizer', None)
+    if tokenizer:
+        new_models['tokenizer'] = tokenizer
+    subset_keys['models'] = [new_models]
+    return subset_keys
+
+
+class AsyncEval(Callback):
+    """Run the eval loop asynchronously as part of a MosaicML platform run.
+
+    This callback is currently experimental. The API may change in the future.
+
+    Args:
+        training_config: Dict[str, Any]: The config from the training run
+        interval: Union[str, int, Time]: The interval describing how often eval runs should be
+            launched. If an integer, it will be assumed to be in :attr:`.TimeUnit.EPOCH`.
+            Otherwise, the unit must be either :attr:`.TimeUnit.EPOCH`, :attr:`.TimeUnit.BATCH`,
+            :attr:`.TimeUnit.TOKEN`, or :attr:`.TimeUnit.SAMPLE`.
+        compute: Optional[Union[ComputeConfig, Dict[str, Any]]]: The compute configuration to
+            use for the eval run. If not provided, the same cluster as the current run and a
+            single, full GPU node will be used.
+    """
+
+    def __init__(
+        self,
+        training_config: Dict[str, Any],
+        interval: Union[str, int, Time],
+        compute: Optional[Union[ComputeConfig, Dict[str, Any]]] = None,
+    ):
+
+        self.training_config = training_config
+
+        if isinstance(interval, str):
+            self.interval = Time.from_timestring(interval)
+        elif isinstance(interval, int):
+            self.interval = Time(interval, TimeUnit.EPOCH)
+        else:
+            self.interval = interval
+
+        self.check_interval = create_interval_scheduler(interval)
+        self.compute = compute
+        self.last_checkpoint: Optional[str] = None
+
+        # Run these during init to fail fast in any of the error cases
+        self.current_run = self._get_current_run()
+        get_eval_parameters(
+            parameters=training_config,
+            checkpoint='test',
+            training_run_name=self.current_run.name,
+        )
+        log.info(
+            f'Initialized AsyncEval callback. Will generate runs at interval {interval}'
+        )
+
+    def run_event(self, event: Event, state: State, logger: Logger) -> None:
```

Comment:
did you check that this waits for the final checkpoint to finish uploading before triggering on `FIT_END`? I believe it would have to run after `RemoteUploaderDownloader` in order for this to hold true. I think this is actually explicitly not true because of the order callbacks are run in (https://github.com/mosaicml/composer/blob/66f412dcc5965d8ea5d5e4d46469893ada06071e/composer/core/engine.py#L278-L294)

Reply:
we could maybe launch the final run on `close`... @eracah got a better way?

Reply:
@dakinggg, it is true that if this runs on fit_end it will run before the RUD fit_end, but this callback seems to operate on locally saved checkpoints, so that means:
1. No need to worry that this callback will run `FIT_END` before RUD does
2. we need a way for it to come after `CheckpointSaver` in the callbacks list
3. We need to figure out a way to maybe edit num_checkpoints_to_keep to be nonzero in order for the callback to work. Otherwise the local checkpoints it needs will be gone!

Reply:
Synced offline, sounds like we'll just want 
1. configure check_interval, so it doesn't run at the last batch_end
2. make a close function in the callback that calls the launch function